File: scala/space-age/src/main/scala/SpaceAge.scala
import scala.language.dynamics

// Wow, scala version's method missing
object SpaceAge extends Dynamic {

  final val PLANET_COEFFICIENT = Map(
    "Mercury" -> 0.2408467,
    "Venus" -> 0.61519726,
    "Mars" -> 1.8808158,
    "Jupiter" -> 11.862615,
    "Saturn" -> 29.447498,
    "Uranus" -> 84.016846,
    "Neptune" -> 164.79132,
    "Earth" -> 1.0
  )

  final val EARTH_SECONDS = 31557600L

  def applyDynamic(name: String)(seconds: Double): Double = {
    val planetName = name.replace("on", "")
    val coefficient = PLANET_COEFFICIENT(planetName)
    seconds / (EARTH_SECONDS * coefficient)
  }
}

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn planet seconds = 
    case planet of
       Mercury -> calcAgeWithCoeffient 0.2408467
       Venus   -> calcAgeWithCoeffient 0.61519726
       Earth   -> calcAgeWithCoeffient 1.0
       Mars    -> calcAgeWithCoeffient 1.8808158
       Jupiter -> calcAgeWithCoeffient 11.862615
       Saturn  -> calcAgeWithCoeffient 29.447498
       Uranus  -> calcAgeWithCoeffient 84.016846
       Neptune -> calcAgeWithCoeffient 164.79132
    where
        earthAge = 31557600
        calcAgeWithCoeffient coeffient = seconds / (earthAge * coeffient)
